src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -0,0 +1,359 @@
+                var semanticDiagnostics = driver.DequeueLocalDiagnostics(analyzer, syntax: false, compilation: compilation);
+                var compilationDiagnostics = driver.DequeueNonLocalDiagnostics(analyzer, compilation);
+
+                lock (_gate)
+                {
+                    if (_completedAnalyzers.Contains(analyzer))
+                    {
+                        // Already stored full analysis result for this analyzer.
+                        continue;
+                    }
+
+                    if (syntaxDiagnostics.Length > 0 || semanticDiagnostics.Length > 0 || compilationDiagnostics.Length > 0 || fullAnalysisResultForAnalyzersInScope)
+                    {
+                        UpdateLocalDiagnostics_NoLock(analyzer, syntaxDiagnostics, fullAnalysisResultForAnalyzersInScope, ref _localSyntaxDiagnosticsOpt);
+                        UpdateLocalDiagnostics_NoLock(analyzer, semanticDiagnostics, fullAnalysisResultForAnalyzersInScope, ref _localSemanticDiagnosticsOpt);
+                        UpdateNonLocalDiagnostics_NoLock(analyzer, compilationDiagnostics, fullAnalysisResultForAnalyzersInScope);
+                    }
+
+                    if (_analyzerExecutionTimeOpt != null)
+                    {
+                        var timeSpan = driver.ResetAnalyzerExecutionTime(analyzer);
+                        _analyzerExecutionTimeOpt[analyzer] = fullAnalysisResultForAnalyzersInScope ?
+                            timeSpan :
+                            _analyzerExecutionTimeOpt[analyzer] + timeSpan;
+                    }
+
+                    if (!_analyzerActionCounts.ContainsKey(analyzer))
+                    {
+                        _analyzerActionCounts.Add(analyzer, getAnalyzerActionCounts(analyzer));
+                    }
+
+                    if (fullAnalysisResultForAnalyzersInScope)
+                    {
+                        _completedAnalyzers.Add(analyzer);
+                    }
+                }
+            }
+        }
+
+        private void UpdateLocalDiagnostics_NoLock(
+            DiagnosticAnalyzer analyzer,
+            ImmutableArray<Diagnostic> diagnostics,
+            bool overwrite,
+            ref Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder>> lazyLocalDiagnostics)
+        {
+            if (diagnostics.IsEmpty)
+            {
+                return;
+            }
+
+            lazyLocalDiagnostics = lazyLocalDiagnostics ?? new Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder>>();
+
+            foreach (var diagsByTree in diagnostics.GroupBy(d => d.Location.SourceTree))
+            {
+                var tree = diagsByTree.Key;
+
+                Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder> allDiagnostics;
+                if (!lazyLocalDiagnostics.TryGetValue(tree, out allDiagnostics))
```

Comment:
I haven't touched most of the code in this file for this PR, but just renamed the file.